src/devices/machine/spifi3.h

```diff
@@ -0,0 +1,465 @@
+	int xfr_phase;
+	int command_pos;
+	bool irq = false;
+	bool drq = false;
+	uint32_t tcounter;
+	uint8_t sync_period;
+	uint8_t clock_conv = 2; // TODO: appropriate value for SPIFI
+	emu_timer *tm;
+	int bus_id;
+
+	// I/O ports
+	devcb_write_line m_irq_handler;
+	devcb_write_line m_drq_handler;
+
+	enum
+	{
+		IDLE
+	};
+
+	enum
+	{
+		// Bus initiated sequences
+		BUSINIT_SETTLE_DELAY = 1,
+		BUSINIT_ASSERT_BUS_SEL,
+		BUSINIT_MSG_OUT,
+		BUSINIT_RECV_BYTE,
+		BUSINIT_ASSERT_BUS_RESEL,
+		BUSINIT_WAIT_REQ,
+		BUSINIT_RECV_BYTE_NACK,
+
+		// Bus SCSI Reset
+		BUSRESET_WAIT_INT,
+		BUSRESET_RESET_BOARD,
+
+		// Disconnected state commands
+		DISC_SEL_ARBITRATION_INIT,
+		DISC_SEL_ARBITRATION,
+		DISC_SEL_ATN_WAIT_REQ,
+		DISC_SEL_ATN_SEND_BYTE,
+		DISC_SEL_WAIT_REQ,
+		DISC_SEL_SEND_BYTE,
+		DISC_REC_ARBITRATION,
+		DISC_REC_MSG_IN,
+		DISC_REC_SEND_BYTE,
+		DISC_RESET,
+
+		// Command sequence
+		CMDSEQ_CMD_PHASE,
+		CMDSEQ_RECV_BYTE,
+
+		// Target commands
+		TARGET_SEND_BYTE,
+		TARGET_CMD_RECV_BYTE,
+		TARGET_MSG_RECV_BYTE,
+		TARGET_MSG_RECV_PAD,
+		TARGET_DISC_SEND_BYTE,
+		TARGET_DISC_MSG_IN,
+		TARGET_DISC_SEND_BYTE_2,
+
+		// Initiator commands
+		INIT_MSG_WAIT_REQ,
+		INIT_XFR,
+		INIT_XFR_SEND_BYTE,
+		INIT_XFR_SEND_PAD_WAIT_REQ,
+		INIT_XFR_SEND_PAD,
+		INIT_XFR_RECV_PAD_WAIT_REQ,
+		INIT_XFR_RECV_PAD,
+		INIT_XFR_RECV_BYTE_ACK,
+		INIT_XFR_RECV_BYTE_NACK,
+		INIT_XFR_FUNCTION_COMPLETE,
+		INIT_XFR_BUS_COMPLETE,
+		INIT_XFR_WAIT_REQ,
+		INIT_CPT_RECV_BYTE_ACK,
+		INIT_CPT_RECV_WAIT_REQ,
+		INIT_CPT_RECV_BYTE_NACK,
+		INIT_XFR_RECV_BYTE_ACK_AUTOMSG
+	};
+
+	enum
+	{
+		// Arbitration
+		ARB_WAIT_BUS_FREE = 1,
+		ARB_COMPLETE,
+		ARB_ASSERT_SEL,
+		ARB_SET_DEST,
+		ARB_RELEASE_BUSY,
+		ARB_TIMEOUT_BUSY,
+		ARB_TIMEOUT_ABORT,
+		ARB_DESKEW_WAIT,
+
+		// Send/receive byte
+		SEND_WAIT_SETTLE,
+		SEND_WAIT_REQ_0,
+		RECV_WAIT_REQ_1,
+		RECV_WAIT_SETTLE,
+		RECV_WAIT_REQ_0
+	};
+
+	enum
+	{
+		STATE_MASK = 0x00ff,
+		SUB_SHIFT = 8,
+		SUB_MASK = 0xff00
+	};
+
+	enum
+	{
+		BUS_BUSY,
+		BUS_FREE_WAIT,
+		BUS_FREE
+	};
+
+	enum dma_direction
+	{
+		DMA_NONE,
+		DMA_IN,
+		DMA_OUT
+	};
+	static inline bool dma_command(dma_direction current_direction)
+	{
+		return current_direction != DMA_NONE;
+	}
+	dma_direction dma_dir;
+
+	// State-related functions
+	void step(bool timeout);
+	void check_irq();
+	void check_drq();
+	void reset_disconnect();
+	void send_byte(scsi_data_target data_source);
+	void recv_byte();
+	void function_complete();
+	void function_bus_complete();
+	void bus_complete();
+	void dma_set(dma_direction dir);
+	void decrement_tcounter(int count = 1);
+	bool transfer_count_zero();
+	void delay(int cycles);
+	void delay_cycles(int cycles);
+	void arbitrate();
+	void clear_fifo();
+	void auto_phase_transfer(int new_phase);
+
+	// AUXCTRL constants and functions
+	const uint32_t AUXCTRL_DMAEDGE = 0x04;
+	const uint32_t AUXCTRL_SETRST = 0x20;
+	const uint32_t AUXCTRL_CRST = 0x40;
+	const uint32_t AUXCTRL_SRST = 0x80;
```

Comment:
Constants should not be instance members.  Also, do all these constants need to be in the header?  Most of them should only be required in the implementation.  Putting excessive amounts of stuff in the header increases the chance of things being recompiled unnecessarily, and also increases compilation time for things that use the class.

Reply:
Done. I moved as many as possible to the cpp files and made them `constexpr` where possible. I use C# at my day job, where `const` is a compile-time constant and implicitly static, so I fell into that pattern when I shouldn't have.